Readme.md

# bitbucket-component

  Proxy component(1) requests to BitBucket, allowing both public and private BitBucket repositories to be used as components.

[![Build Status](https://travis-ci.org/stephenmathieson/bitbucket-component.png)](https://travis-ci.org/stephenmathieson/bitbucket-component)

## Installation

    $ npm install -g stephenmathieson/bitbucket-component

## Setup

  bitbucket-component(1) expects an `{env}.config.json` to exist in your current directory.  Put your port number, username and password in there.

  **env.config.json**

```json
{
  "port": 1234,
  "username": "...",
  "password": "..."
}
```

  Switch environments by setting `NODE_ENV` before running `bitbucket-component`.

## License 

(The MIT License)

Copyright (c) 2013 Stephen Mathieson &lt;me@stephenmathieson.com&gt;

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

History.md

1.0.1 / 2014-04-04
==================

 * Fix config.json path when using binary
 * Remove old binary, replace with new :)

1.0.0 / 2014-04-04
==================

 * Refactor to leverage BitBucket's new "raw" API

0.0.1 / 2010-01-03
==================

  * Initial release


test/index.js

var app = require('..');
var request = require('supertest');

describe('app', function () {
  before(function (done) {
    app.listen(5000, done);
  });

  after(function (done) {
    app.close(done);
  });

  describe('POST /', function () {
    it('should 405', function (done) {
      request(app)
      .post('/')
      .expect(405, done);
    });
  });

  describe('GET /', function () {
    it('should 404', function (done) {
      request(app)
      .get('/')
      .expect(404, done);
    });
  });

  describe('GET /owner', function () {
    it('should 404', function (done) {
      request(app)
      .get('/owner')
      .expect(404, done);
    });
  });

  describe('GET /owner/name', function () {
    it('should 404', function (done) {
      request(app)
      .get('/owner/name')
      .expect(404, done);
    });
  });

  describe('GET /owner/name/version', function () {
    it('should 404', function (done) {
      request(app)
      .get('/owner/name/version')
      .expect(404, done);
    });
  });

  describe('GET /stephenmathieson/nope/4.5.6/file.js', function () {
    it('should 404', function (done) {
      request(app)
      .get('/stephenmathieson/node/4.5.6/file.js')
      .expect(404, done);
    });
  });

  describe('GET /stephenmathieson/testything/0.0.4/component.json', function () {
    it('should 200', function (done) {
      request(app)
      .get('/stephenmathieson/testything/0.0.4/component.json')
      .expect(200, done);
    });

    it('should set headers', function (done) {
      request(app)
      .get('/stephenmathieson/testything/0.0.4/component.json')
      .expect('content-type', /.+/)
      .expect('content-length', /.+/)
      .expect('last-modified', /.+/, done);
    });

    it('should provide component.json', function (done) {
      var json = {
        name: 'testything',
        repo: 'stephenmathieson/testything',
        description: 'a testy thing',
        version: '0.0.4',
        keywords: [],
        dependencies: {},
        development: {},
        license: 'MIT',
        main: 'index.js',
        scripts: ['index.js', 'lib/apples.js'],
        styles: ['testything.css']
      };
      var str = JSON.stringify(json, null, 2);

      request(app)
      .get('/stephenmathieson/testything/0.0.4/component.json')
      .expect(str + '\n', done);
    });
  });

  describe('GET /stephenmathieson/testything/0.0.4/lib/apples.js', function () {
    it('should 200', function (done) {
      request(app)
      .get('/stephenmathieson/testything/0.0.4/lib/apples.js')
      .expect(200, done);
    });

    it('should set headers', function (done) {
      request(app)
      .get('/stephenmathieson/testything/0.0.4/lib/apples.js')
      .expect('content-type', /.+/)
      .expect('content-length', /.+/)
      .expect('last-modified', /.+/, done);
    });

    it('should provide lib/apples.js', function (done) {
      var str = '\nmodule.exports = \'apples\'';

      request(app)
      .get('/stephenmathieson/testything/0.0.4/lib/apples.js')
      .expect(str + '\n', done);
    });
  });
});



test/acceptance/gh-6-config.js

var spawn = require('win-fork'),
    path = require('path'),
    fs = require('fs'),
    request = require('supertest'),
    rimraf = require('rimraf');

var bin = path.join(__dirname, '../../bin/server.js'),
    conf = path.join(__dirname, 'gh-6-config.json'),
    repos = path.join(__dirname, 'gh-6-repos');


describe('bitbucket-component(1)', function () {

  var server;

  before(function (done) {

    // write config file
    fs.writeFileSync(conf, JSON.stringify({
      directory: repos,
      port: 5000
    }, null, 2));

    // start server
    server = spawn(bin, [ '-c', conf, '--verbose' ]);

    server
      .stdout
      // assuming the first data received is
      // the "im listening" message
      .once('data', function () {
        done();
      });
  });

  after(function () {
    fs.unlinkSync(conf);
    rimraf.sync(repos);
    process.kill(server);
  });

  it('should support a config file', function (done) {

    var file = '/stephenmathieson/testything/0.0.4/component.json';

    request('http://localhost:5000')
      .get(file)
      .expect(200, function (err) {
        if (err) return done(err);
        fs.existsSync(repos).should.be.true;
        fs.existsSync(path.join(repos, file)).should.be.true;
        done();
      });
  });
});


lib/bitbucket.js

'use strict';

var fs = require('fs'),
    path = require('path'),
    tar = require('tar'),
    zlib = require('zlib'),
    mkdirp = require('mkdirp'),
    https = require('https'),
    errors = require('http-error'),
    os = require('os'),
    once = require('once');


/**
 * Download a tarball from `opts.url`, parsing out `owner`,
 * `name`, and `version`.  Invokes `cb(err, directory_to_untarred_repo)`.
 *
 * @api private
 * @param {Options} opts
 * @param {String} owner
 * @param {String} name
 * @param {String} version
 * @param {Function} cb
 */
exports.get = function (opts, owner, name, version, cb) {
  // only fire once
  cb = once(cb);

  var url = opts.url
        .replace('{owner}', owner)
        .replace('{name}', name)
        .replace('{version}', version);

  var temp = path.join(os.tmpDir(), owner + '-' + name + '-' + version);

  var dir = path.join(opts.directory, owner, name, version);


  /**
   * Callback for completion.  Moves the extracted tarball
   * to its desired location (`{directory}/owner/component/version`).
   *
   * Added complexity as the tarball directly from bitbucket
   * contains an arbitrary directory.
   *
   * @api private
   * @param {http.IncomingMessage} res
   */
  function onEnd(res) {
    var attachment = path.join(temp, exports.attachmentName(res));

    mkdirp(dir, function (err) {
      if (err) return cb(err);

      fs.rename(attachment.split('.tar.gz')[0], dir, function (err) {
        if (err) return cb(err);

        cb(null, dir);
      });
    });
  }

  /**
   * Callback for errors.  Adds useful properties to
   * the given error, allowing for easier debugging.
   *
   * @api private
   * @param {Error} err
   */
  function error(err) {
    err.repo = owner + '/' + name;
    err.version = version;
    cb(err);
  }

  /**
   * Handle the given `res`
   *
   * @api private
   * @param {http.IncomingMessage} res
   */
  function handleResponse(res) {

    // 400s and 500s
    if ((res.statusCode / 100 | 0) > 3) {
      var err = new errors[res.statusCode]('failed to fetch ' + url);
      return cb(error(err));
    }

    res.on('error', error);
    res.pipe(zlib.createGunzip())
      .on('error', error)
      .pipe(tar.Extract(temp))
        .on('error', error)
        .on('end', onEnd.bind(null, res));
  }

  // don't re-download the tarball if it already exists
  fs.exists(dir, function (exists) {
    if (exists) return cb(null, dir);
    https.get(url, handleResponse).on('error', error);
  });
};

/**
 * Parse the attacement name from the given `res`
 *
 * @api private
 * @param {http.IncomingMessage} res
 * @return {String}
 */
exports.attachmentName = function (res) {
  if (!res.headers || !res.headers['content-disposition']) return null;
  return res
    .headers['content-disposition']
    .split('attachment; filename=')[1];
};


lib/bitbucket-component.js

'use strict';

var bitbucket = require('./bitbucket'),
    express = require('express'),
    path = require('path'),
    merge = require('util')._extend;


/**
 * Create an express app for serving components from bitbucket.org
 *
 * Options:
 *
 *   - `directory` the directory to download repositories in
 *   - `password` bitbucket password
 *   - `username` bitbucket username
 *   - `maxAge` max age for file cache; see http://expressjs.com/api.html#res.sendfile
 *
 * @api public
 * @param {Object} opts
 * @return {Function}
 */

exports = module.exports = function (opts) {
  opts = merge({
    directory: null,
    username: null,
    password: null,
    maxAge: 0
  }, opts);

  if (opts.password && !opts.username) throw new Error('must provide a username');
  if (opts.username && !opts.password) throw new Error('must provide a password');

  opts.url = opts.username
    ? 'https://' + opts.username + ':' + opts.password
      + '@bitbucket.org/{owner}/{name}/get/{version}.tar.gz'
    : 'https://bitbucket.org/{owner}/{name}/get/{version}.tar.gz';

  var app = express();

  app.get('/:owner/:name/:version/:file(*)', function (req, res, next) {
    var owner = req.param('owner'),
        name = req.param('name'),
        version = req.param('version'),
        file = req.param('file');

    bitbucket.get(opts, owner, name, version, function (err, directory) {
      if (err) return next(err);
      res.sendfile(path.join(directory, file), { maxAge: opts.maxAge });
    });
  });

  return app;
};

exports.bitbucket = bitbucket;


app.js
#!/usr/bin/env node

/**
 * Module dependencies.
 */

var http = require('http');
var request = require('hyperquest');
var url = require('url');
var ms = require('ms');

/**
 * Conf.
 */

var pkg = require('./package');
var config = module.parent
  ? require('./config')
  : require(process.cwd() + '/config');

// build host string based on credentials
var host = config.username && config.password
  ? 'https://' + config.username + ':' + config.password
    + '@' + 'bitbucket.org'
  : 'https://bitbucket.org';

/**
 * Expose our app.
 */

var app = module.exports = http.createServer();

// $ node app.js
if (!module.parent) {
  app.listen(config.port, function () {
    console.log();
    console.log('  bitbucket/component proxy v%s', pkg.version);
    console.log();
    console.log('    listening on port %d', config.port);
    console.log();
  });
}

// logger (disabled when testing)
if ('test' != config.env) {
  app.on('request', function (req, res) {
    var start = new Date;
    res.once('finish', function () {
      var end = new Date;
      var str =
          req.method
        + ' '
        + req.url
        + ' '
        + res.statusCode
        + ' ('
        + ms(end - start)
        + ')';
      console.log(str);
    });
  });
}

// "main" request handler
app.on('request', function (req, res) {
  if ('GET' != req.method) return methodNotSupported(req, res);

  var component = parse(req);

  if (!component) {
    return notFound(req, res);
  }

  // https://{host}/{owner}/{name}/raw/{version}/{file}
  var url = host
    + '/'
    + component.author
    + '/'
    + component.name
    + '/raw/'
    + component.version
    + '/'
    + component.file;

  request(url)
  .once('response', function (response) {
    if (404 == response.statusCode) return notFound(req, res);

    // set relevant header fields
    var headers = response.headers;
    var relevant = [
      'content-type', 'content-length', 'last-modified'
    ];
    for (var i = relevant.length - 1; i >= 0; i--) {
      var header = relevant[i];
      if (headers[header]) {
        res.setHeader(header, headers[header]);
      }
    }

    res.statusCode = response.statusCode;
  })
  .on('error', interalError.bind(null, req, res))
  .pipe(res);
});

/**
 * Parse the requested URL for a matching component.
 *
 * @api private
 * @param {http.IncomingMessage} req
 * @return {Object}
 */

function parse(req) {
  var parsed = url.parse(req.url);
  var parts = parsed.pathname.split('/').filter(Boolean);
  if (4 > parts.length) return;
  return {
    author: parts[0],
    name: parts[1],
    version: parts[2],
    file: parts.slice(3).join('/')
  };
}

/**
 * 405 callback.
 */

function methodNotSupported(req, res) {
  res.statusCode = 405;
  res.end(req.method + ' not supported.');
}

/**
 * 500 callback.
 */

function interalError(req, res, err) {
  res.statusCode = 500;
  res.end(err.message);
}

/**
 * 404 callback.
 */

function notFound(req, res) {
  res.statusCode = 404;
  res.end('not found');
}
